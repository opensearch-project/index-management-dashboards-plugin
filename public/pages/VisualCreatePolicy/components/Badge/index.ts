/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import Badge from "./Badge";

export default Badge;
