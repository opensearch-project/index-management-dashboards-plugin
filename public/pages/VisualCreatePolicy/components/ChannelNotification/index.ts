/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ChannelNotification from "./ChannelNotification";

export default ChannelNotification;
