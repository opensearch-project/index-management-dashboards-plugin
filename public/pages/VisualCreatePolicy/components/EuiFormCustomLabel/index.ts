/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import EuiFormCustomLabel from "./EuiFormCustomLabel";

export default EuiFormCustomLabel;
