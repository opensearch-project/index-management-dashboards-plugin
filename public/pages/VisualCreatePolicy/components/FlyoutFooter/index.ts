/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import FlyoutFooter from "./FlyoutFooter";

export default FlyoutFooter;
