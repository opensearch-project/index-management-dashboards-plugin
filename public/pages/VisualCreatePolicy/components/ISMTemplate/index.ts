/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ISMTemplate from "./ISMTemplate";

export default ISMTemplate;
