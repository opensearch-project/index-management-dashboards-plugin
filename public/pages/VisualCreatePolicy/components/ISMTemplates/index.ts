/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import ISMTemplates from "./ISMTemplates";

export default ISMTemplates;
