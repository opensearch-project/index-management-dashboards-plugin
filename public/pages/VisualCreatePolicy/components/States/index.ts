/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import States from "./States";

export default States;
