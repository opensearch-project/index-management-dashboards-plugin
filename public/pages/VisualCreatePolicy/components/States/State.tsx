/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { EuiAccordion, EuiText, EuiPanel, EuiFlexGroup, EuiFlexItem, EuiButtonIcon, EuiToolTip } from "@elastic/eui";
import "brace/theme/github";
import "brace/mode/json";
import { State as StateData } from "../../../../../models/interfaces";
import { ModalConsumer } from "../../../../components/Modal";
import ConfirmationModal from "../../../../components/ConfirmationModal";
import Badge from "../Badge";
import TransitionContent from "../Transition/TransitionContent";
import { makeId } from "../../../../utils/helpers";
import { actionRepoSingleton } from "../../utils/helpers";

interface StateProps {
  state: StateData;
  isInitialState: boolean;
  idx: number;
  onClickEditState: (state: StateData) => void;
  onClickDeleteState: (idx: number) => void;
  isReadOnly: boolean;
}

const State = ({ state, isInitialState, idx, onClickEditState, onClickDeleteState, isReadOnly = false }: StateProps) => (
  <EuiAccordion
    style={{ padding: "15px" }}
    id={state.name}
    buttonClassName="state-accordion"
    buttonContent={
      <EuiFlexGroup justifyContent="center" alignItems="center">
        <EuiFlexItem grow={false}>
          <EuiText>
            <strong>{state.name}</strong>
          </EuiText>
        </EuiFlexItem>
        {isInitialState && (
          <EuiFlexItem grow={false}>
            <EuiPanel paddingSize="none">
              <EuiText size="xs" style={{ padding: "0px 10px" }}>
                Initial state
              </EuiText>
            </EuiPanel>
          </EuiFlexItem>
        )}
        {!!state.actions?.length && (
          <EuiFlexItem grow={false}>
            <Badge text="Actions" number={state.actions.length} />
          </EuiFlexItem>
        )}
        {!!state.transitions?.length && (
          <EuiFlexItem grow={false}>
            <Badge text="Transitions" number={state.transitions.length} />
          </EuiFlexItem>
        )}
      </EuiFlexGroup>
    }
    extraAction={
      !isReadOnly && (
        <EuiFlexGroup gutterSize="m">
          <EuiFlexItem grow={false}>
            <ModalConsumer>
              {({ onShow, onClose }) => (
                <EuiToolTip position="top" content={<p>Delete state</p>}>
                  <EuiButtonIcon
                    iconType="trash"
                    aria-label="Delete"
                    color="danger"
                    onClick={() =>
                      onShow(ConfirmationModal, {
                        title: "Delete state",
                        bodyMessage: (
                          <EuiText>
                            <span>
                              Delete &quot;<strong>{state.name}</strong>&quot; permanently? Deleting the state will result in deleting all
                              transitions.
                            </span>
                          </EuiText>
                        ),
                        actionMessage: "Delete state",
                        actionProps: { color: "danger" },
                        modalProps: { maxWidth: 600 },
                        onAction: () => onClickDeleteState(idx),
                        onClose,
                      })
                    }
                    data-test-subj="state-delete-button"
                  />
                </EuiToolTip>
              )}
            </ModalConsumer>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiToolTip position="top" content={<p>Edit state</p>}>
              <EuiButtonIcon iconType="pencil" aria-label="Edit" color="primary" onClick={() => onClickEditState(state)} />
            </EuiToolTip>
          </EuiFlexItem>
        </EuiFlexGroup>
      )
    }
    paddingSize="l"
  >
    <EuiFlexGroup direction="column" gutterSize="l">
      <EuiFlexItem grow={false}>
        <EuiText>
          <h4>Actions</h4>
        </EuiText>
      </EuiFlexItem>
      <EuiFlexItem>
        {!state.actions?.length ? (
          <EuiText>No actions. Edit state to add actions.</EuiText>
        ) : (
          <EuiFlexGroup wrap>
            {state.actions.map((action, index) => (
              <EuiFlexItem grow={false} key={`${makeId()}-${index}`}>
                <EuiPanel>{actionRepoSingleton.getUIActionFromData(action).content()}</EuiPanel>
              </EuiFlexItem>
            ))}
          </EuiFlexGroup>
        )}
      </EuiFlexItem>
      <EuiFlexItem grow={false}>
        <EuiText>
          <h4>Transitions</h4>
        </EuiText>
      </EuiFlexItem>
      <EuiFlexItem>
        {!state.transitions?.length ? (
          <EuiText>No transitions. Edit state to add transitions.</EuiText>
        ) : (
          <EuiFlexGroup wrap>
            {state.transitions.map((transition, index) => (
              <EuiFlexItem grow={false} key={`${makeId()}-${index}`}>
                <EuiPanel>
                  <TransitionContent transition={transition} />
                </EuiPanel>
              </EuiFlexItem>
            ))}
          </EuiFlexGroup>
        )}
      </EuiFlexItem>
    </EuiFlexGroup>
  </EuiAccordion>
);

export default State;
