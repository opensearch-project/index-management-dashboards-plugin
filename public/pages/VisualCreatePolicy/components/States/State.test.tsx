/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { screen, render } from "@testing-library/react";
import { fireEvent, waitFor } from "@testing-library/dom";
import State from "./State";
import { State as StateData } from "../../../../../models/interfaces";
import { ModalProvider, ModalRoot } from "../../../../components/Modal";
import { ServicesConsumer, ServicesContext } from "../../../../services";
import { browserServicesMock } from "../../../../../test/mocks";

describe("<State /> spec", () => {
  it("renders the component", () => {
    const state: StateData = {
      name: "some_name",
      actions: [{ close: {} }, { open: {} }, { delete: {} }],
      transitions: [{ state_name: "elsewhere", conditions: {} }],
    };
    const { container } = render(
      <State state={state} isInitialState={true} idx={2} onClickEditState={() => {}} onClickDeleteState={() => {}} />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it("does not show initial state when it is not the initial state", async () => {
    const state: StateData = {
      name: "some_name",
      actions: [{ close: {} }, { open: {} }, { delete: {} }],
      transitions: [{ state_name: "elsewhere", conditions: {} }],
    };
    render(<State state={state} isInitialState={false} idx={2} onClickEditState={() => {}} onClickDeleteState={() => {}} />);

    await waitFor(() => screen.getByText("some_name", { exact: false }));

    expect(screen.queryByText("Initial state", { exact: false })).toBeNull();
  });

  it("shows no transitions message and no actions message if none defined", async () => {
    const state: StateData = {
      name: "some_name",
      actions: [],
      transitions: [],
    };
    render(<State state={state} isInitialState={false} idx={2} onClickEditState={() => {}} onClickDeleteState={() => {}} />);

    await waitFor(() => screen.getByText("some_name", { exact: false }));

    expect(screen.queryByText("No transitions. Edit state to add transitions.", { exact: false })).not.toBeNull();
    expect(screen.queryByText("No actions. Edit state to add actions.", { exact: false })).not.toBeNull();
  });

  it("renders delete modal when deleting a state", async () => {
    const state: StateData = {
      name: "some_name",
      actions: [{ close: {} }, { open: {} }, { delete: {} }],
      transitions: [{ state_name: "elsewhere", conditions: {} }],
    };
    const onClickDeleteState = jest.fn();
    const { getByTestId } = render(
      <ServicesContext.Provider value={browserServicesMock}>
        <ModalProvider>
          <ServicesConsumer>{(services) => services && <ModalRoot services={services} />}</ServicesConsumer>
          <State state={state} isInitialState={true} idx={2} onClickEditState={() => {}} onClickDeleteState={onClickDeleteState} />
        </ModalProvider>
      </ServicesContext.Provider>
    );

    fireEvent.click(getByTestId("state-delete-button"));

    await waitFor(() => screen.getByText("Deleting the state will result in deleting all transitions.", { exact: false }));

    fireEvent.click(getByTestId("confirmationModalActionButton"));

    expect(onClickDeleteState).toHaveBeenCalled();
  });
});
