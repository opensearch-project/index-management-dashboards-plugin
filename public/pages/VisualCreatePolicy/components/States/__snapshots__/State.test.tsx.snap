// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<State /> spec renders the component 1`] = `
<div
  class="euiAccordion"
  style="padding: 15px;"
>
  <div
    class="euiAccordion__triggerWrapper"
  >
    <button
      aria-controls="some_name"
      aria-expanded="false"
      class="euiAccordion__button state-accordion"
      id="some_html_id"
      type="button"
    >
      <span
        class="euiAccordion__iconWrapper"
      >
        EuiIconMock
      </span>
      <span
        class="euiIEFlexWrapFix"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentCenter euiFlexGroup--directionRow euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiText euiText--medium"
            >
              <strong>
                some_name
              </strong>
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiPanel euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
            >
              <div
                class="euiText euiText--extraSmall"
                style="padding: 0px 10px;"
              >
                Initial state
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <div
                class="euiTextAlign euiTextAlign--center"
              >
                <span
                  class="euiTextColor euiTextColor--subdued"
                >
                  <p>
                    Actions
                     
                    <span
                      class="euiNotificationBadge euiNotificationBadge--subdued"
                    >
                      3
                    </span>
                  </p>
                </span>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <div
                class="euiTextAlign euiTextAlign--center"
              >
                <span
                  class="euiTextColor euiTextColor--subdued"
                >
                  <p>
                    Transitions
                     
                    <span
                      class="euiNotificationBadge euiNotificationBadge--subdued"
                    >
                      1
                    </span>
                  </p>
                </span>
              </div>
            </div>
          </div>
        </div>
      </span>
    </button>
    <div
      class="euiAccordion__optionalAction"
    >
      <div
        class="euiFlexGroup euiFlexGroup--gutterMedium euiFlexGroup--directionRow euiFlexGroup--responsive"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <span
            class="euiToolTipAnchor"
          >
            <button
              aria-label="Delete"
              class="euiButtonIcon euiButtonIcon--danger euiButtonIcon--empty euiButtonIcon--xSmall"
              data-test-subj="state-delete-button"
              type="button"
            >
              EuiIconMock
            </button>
          </span>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrowZero"
        >
          <span
            class="euiToolTipAnchor"
          >
            <button
              aria-label="Edit"
              class="euiButtonIcon euiButtonIcon--primary euiButtonIcon--empty euiButtonIcon--xSmall"
              type="button"
            >
              EuiIconMock
            </button>
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    aria-labelledby="some_html_id"
    class="euiAccordion__childWrapper"
    id="some_name"
    role="region"
    tabindex="-1"
  >
    <div>
      <div
        class="euiAccordion__padding--l"
      >
        <div
          class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionColumn euiFlexGroup--responsive"
        >
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiText euiText--medium"
            >
              <h4>
                Actions
              </h4>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
                >
                  Close
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
                >
                  Open
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
                >
                  Delete
                </div>
              </div>
            </div>
          </div>
          <div
            class="euiFlexItem euiFlexItem--flexGrowZero"
          >
            <div
              class="euiText euiText--medium"
            >
              <h4>
                Transitions
              </h4>
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap"
            >
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <div
                  class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
                >
                  <div
                    class="euiText euiText--medium"
                  >
                    <dl>
                      <dt>
                        Destination state
                      </dt>
                      <dd>
                        elsewhere
                      </dd>
                      <dt>
                        Transition trigger logic
                      </dt>
                      <dd>
                        No condition
                      </dd>
                    </dl>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
